Support for Multiple component analysis

xispa opened the pull request:
## Description of the issue/feature this PR addresses

This Pull Request adds the Multiple Component analysis. Spectroscopic methods are common analytical methods for which multiple component analysis is used, like are Atomic Absorption Spectroscopy (AAS), Inductively Coupled Plasma Spectroscopy (ICP) or Ultraviolet and Visible (UV). Basically, the concentration of individual components/analytes is detected simultaneously, with a single analyzer.

Creation of an Multicomponent Analysis Service:

![Captura de 2022-09-04 12-36-32](https://user-images.githubusercontent.com/832627/188313803-0e8cf374-baf3-462a-ac66-9137047e9551.png)

Selection of analytes on Sample creation:

![Captura de 2022-09-04 21-46-15](https://user-images.githubusercontent.com/832627/188331110-38caf0b3-373c-4b73-85fa-718283fe6a93.png)


Multicomponent analyses in worksheet:

![Captura de 2022-09-04 12-36-47](https://user-images.githubusercontent.com/832627/188313830-285775cd-8e3a-48bc-b7b1-d527b9e9dcff.png)


Multicomponent analyses in sample:

![Captura de 2022-09-04 12-37-38](https://user-images.githubusercontent.com/832627/188313839-6da1f147-cb8b-4d80-9581-816f692d5278.png)



## Current behavior before PR

SENAITE does not support multi-component analysis. Lab is forced to create an individual Analysis Service for every analyte/component

## Desired behavior after PR is merged

SENAITE does support multi-component analysis. Analytes can be defined as components of a single Analysis Service

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

xispa committed:
Add MultiComponent and Analytes fields to AnalysisService type

xispa committed:
Simplify validator

xispa committed:
Add new func for the creation of a copy from a service

xispa committed:
Validate Service's Analytes on creation/edition

xispa committed:
Add skip parameter to copy_service func

xispa committed:
Automatic creation of analytes (as Analysis objects) from Multicomponent analyses

xispa committed:
Add fields for Analytes retrieval from Analysis content type

xispa committed:
Add isAnalyte index in analyses catalog

xispa committed:
Less intrusive addition of isAnalyte index

xispa committed:
Display analytes as children in analyses listing

xispa committed:
Auto-set result to MultiComponent on Analyte results capture

xispa committed:
Do not allow to submit a multi-component unless all analytes submitted

xispa committed:
Assign method and instrument from multi-component to analytes

xispa committed:
Bad spelling

xispa committed:
Do not display the method in analytes

xispa committed:
Prevent max recursion depth

xispa committed:
Rejection of Multi-component causes the rejection of all analytes

xispa committed:
Do not allow to retract analytes, but multi-component only

xispa committed:
Create new analytes when a multi-component analysis is retracted

xispa committed:
Simplify and cleanup

xispa committed:
Rely on IRetracted marker interface on guard_retract

xispa committed:
Cleanup

xispa committed:
Cleanup

xispa committed:
Retraction is only permitted for multi-analysis, but when analytes suit

xispa committed:
Auto-reject multi-component when all its analytes are rejected

xispa committed:
Auto-verify analytes when multi-component is verified

xispa committed:
Do not display the result of a multi-component analysis in listing

xispa committed:
Do not display the Submitter for multi-component analyses in listings

xispa committed:
Allow to retest a multi-component analysis

xispa committed:
Cleanup

xispa committed:
Automatically assign analytes when multi-component is assigned

xispa committed:
Automatically unassign analytes when a multi-analysis is unassigned

xispa committed:
Do not display analytes in the Add analyses view from inside Worksheet

xispa committed:
Render analytes as a flat list in worksheet results entry

xispa committed:
Set the analysis' default method on creation

xispa committed:
Performance: enhance retrieval of methods from analysis/service

xispa committed:
Optimize and clean-up accessors from analysis/service types

xispa committed:
Added ico in services listing next to multi-component services

xispa committed:
Fix retract doctests

xispa committed:
Fixture for tests that do not have a regular request

xispa committed:
Add doctest for multi-analysis/service

xispa committed:
Changelog

xispa committed:
Fix retest guard

xispa committed:
Complete doctest

xispa committed:
Fix bad message in test

xispa committed:
Restore the check against calculation interim fields when validating service keywords

xispa committed:
Added Assignment section in multicomponent doctest

xispa committed:
Remove unused import (F401 'bika.lims.workflow as wf' imported but unused)

xispa committed:
Allow to select Analytes on sample registration

xispa committed:
Do not display the "Hidden" checkbox for multi-component analyses

ramonski commented on src/bika/lims/api/analysisservice.py:
```diff
@@ -153,3 +160,116 @@ def get_service_dependencies_for(service):
+    fields = api.get_fields(service).values()
+    for field in fields:
+        if skip_field(field):
+            continue
```
Please note that you are iterating here for every field through the list of skip fields. Better would be here a dictionary, e.g.:
`skip.get(field.getName(), False)` or `skip.get(field.getType(), False)`

https://www.freecodecamp.org/news/big-o-notation-why-it-matters-and-why-it-doesnt-1674cfa8a23c/

xispa replied:
Thanks for the heads-up. Given the fact that copying objects is something we do quite often, I made it better and created a function in the generic api with https://github.com/senaite/senaite.core/pull/2120/commits/5e487a80ec98fdfc1a025bb3632c350d141cc9d9

ramonski commented on src/bika/lims/api/analysisservice.py:
```diff
@@ -153,3 +160,116 @@ def get_service_dependencies_for(service):
+    brains = get_by_keyword(keyword)
+    if instance:
+        uid = api.get_uid(instance)
+        brains = filter(lambda brain: api.get_uid(brain) != uid, brains)
```
Can't we simply check here `len(brains) > 1`?

xispa replied:
:thinking: you are right, no need to check by uid. Resolved with https://github.com/senaite/senaite.core/pull/2120/commits/fc81a353a4154e99f477f80b9fbd36692e755676

ramonski commented on src/bika/lims/browser/fields/aranalysesfield.py:
```diff
@@ -274,11 +275,13 @@ def add_analysis(self, instance, service, **kwargs):
 
         if not analyses:
             # Create the analysis
-            new_id = self.generate_analysis_id(instance, service)
```
Why is this no longer needed?

xispa replied:
Because it was non-sense to always do this dance before creating or copying an analysis. So I simply [moved this logic into the `create_analysis` function](https://github.com/senaite/senaite.core/pull/2120/files#diff-7e06c4886a87d3378dd9a03774e3a66f5d8350e6b9de28c8b815b439d0bacf79R77-R81) and removed the generation of the id before analysis creation from here and from [other places as well](https://github.com/senaite/senaite.core/pull/2120/files#diff-7e06c4886a87d3378dd9a03774e3a66f5d8350e6b9de28c8b815b439d0bacf79L404-L409)

ramonski commented on src/bika/lims/validators.py:
```diff
@@ -252,52 +253,10 @@ class ServiceKeywordValidator:
+        err_msg = serviceapi.check_keyword(value, instance)
+        if err_msg:
+            ts = api.get_tool("translation_service")
+            return to_utf8(ts.translate(_(err_msg)))
```
I think the error messages from the service api `check_keyword` need to be wrapped into the message factory `_(...)`. Otherwise, they won't be grabbed for translation.

xispa replied:
Resolved with https://github.com/senaite/senaite.core/pull/2120/commits/199ef66a82a1ce1750aac23076c999cf6eee579a

ramonski commented on src/bika/lims/api/analysisservice.py:
```diff
@@ -153,3 +160,116 @@ def get_service_dependencies_for(service):
+
+    # Ensure the format is valid
+    if re.findall(RX_SERVICE_KEYWORD, keyword):
+        return "Validation failed: keyword contains invalid characters"
```
Wrap into message factory `_(...)`

xispa replied:
Resolved with https://github.com/senaite/senaite.core/pull/2120/commits/199ef66a82a1ce1750aac23076c999cf6eee579a

ramonski commented on src/senaite/core/upgrade/v02_03_000.py:
```diff
@@ -411,3 +421,71 @@ def fixed_point_value_to_string(value, precision):
+    """
+    logger.info("Applying IRetracted interface to retracted analyses ...")
+    query = {
+        "portal_type": ["Analysis", "ReferenceAnalysis", "DuplicateAnalysis"]
```
Please search here for `"review_state": "retracted"`, or for both `"review_state": ["retracted", "rejected"]` and merge with the function below.

ramonski replied:
I got now why you did the upgrade step like this, because the current state can be different due to re-activation and you would like to mark any analyses that have ever been rejected / retracted.
However, I think we should do a `noLonverProvides` if this happened, because it does not make sense to keep it, because these are then no longer rejected/retracted.

xispa replied:
Yes, I did not filter by review state because I don't know if there are analyses in a further status passed retracted/rejected. I thought the same, that shouldn't make sense. 

However, I later I realized that client-specific add-ons (not my case) may exist that do consider additional states past reject/retract, specially after the "dispacth" transition and/or senaite.storage were added. And I left the query w/o changes.

Still, filtering by review-state won't have any impact on any of the instances using the add-ons I am aware of.

ramonski replied:
Yes, that dispatch state...

Not sure if it was a good idea, but it might make sense when using senaite.referral or when subcontracting external labs for analysis.

Nevertheless, I would change the upgrade step to either iterate once over all analyses or consider only final rejected/retracted states.

The reason is that we have here instances with 500k Analyses and it is a pain to iterate through all of them for migration (takes 2.5 days currently)

xispa replied:
Sure, let's keep things simpler and faster https://github.com/senaite/senaite.core/pull/2120/commits/81aa786454859b8a0795c4b69d2a58a23ee9f487

xispa committed:
Added copy_object func into the API

xispa committed:
Leave ShortTitle field empty when creating a copy

xispa committed:
Simplify keyword check (no need to filter by uid)

xispa committed:
Wrap validator messages with messageFactory

xispa committed:
Assume "reject" and "retract" are final statuses

ramonski commented:
Creating a new Analysis Service with Analytes complains about that the Keyword is already in use, no matter which keyword is chosen:

<img width="977" src="https://user-images.githubusercontent.com/713193/189062731-c40a171e-3f84-48b2-9127-a3ebff9ae67c.png">

<img width="1736" src="https://user-images.githubusercontent.com/713193/189062824-35d2a411-f59a-46aa-822f-32f37c818528.png">

However, although the error is displayed, the Analysis service is created with that keyword:

<img width="1727" src="https://user-images.githubusercontent.com/713193/189063676-f6fd795d-d9fa-4f09-9cb1-655a0d6962bf.png">

And interestingly, it is also updated when the keyword changes (which also causes the error message):

<img width="1730" src="https://user-images.githubusercontent.com/713193/189063346-17b1b635-5c1c-4734-bb25-646cf2ddcfcc.png">

ramonski commented:
Just a general thought:

You implemented now Analytes as siblings to existing analyses, but I think they should be actually contained inside analyses.
I know that we did it the same for sample partitions, but I think that decision should be also rethought...

xispa committed:
Do not show "Existing keyword" error when creating an Analysis Service

xispa commented:
> Creating a new Analysis Service with Analytes complains about that the Keyword is already in use, no matter which keyword is chosen:
> [...]
> However, although the error is displayed, the Analysis service is created with that keyword

Addressed with https://github.com/senaite/senaite.core/pull/2120/commits/245354b974a7175b7949ed78f19288bfdddc15db

xispa commented:
> Just a general thought:
> 
> You implemented now Analytes as siblings to existing analyses, but I think they should be actually contained inside analyses. I know that we did it the same for sample partitions, but I think that decision should be also rethought...

Yes, it might make sense. However, making the system work with Analysis objects that are not contained in AnalysisRequest entails a lot of additional changes, not only in ARAnalysisField (plenty of `objectValues('Analysis')` in there, but also in logic that is related with Worksheet. The good thing about this approach is that since analytes become regular "Analyses", the existing logic works without any changes other than those related with visualization and guards. I think this discussion will come up again when we move to DX and need to rewrite `ARAnalysesField`. Maybe this is something we could reconsider by then.

ramonski committed:
Merge branch '2.x' into multi-analyses

ramonski commented:
Some more findings from integration testing:

- Analytes that have the same keyword as existing Analysis services, e.g. `Zn` collide in Worksheets and show only one of both
- Transposed worksheets show the Multi Component Analysis in between the Analytes and not on top. Maybe we need to fix the sort key on the fly there?

ramonski commented:
Assigning a calculation to a multi component disables all result fields of the Analytes.

ramonski commented:
It is not possible to retract an Analyte result that is in `to_be_verified` state

ramonski commented:
When an Analyte has the same keyword as an Analysis, e.g. `Zn`, and the Analysis exists together in the same Sample with the Analyte, it is not possible to remove the Analysis anymore over the "Manage Analyses" view.

ramonski commented:
All Analytes are added when a sample is copied, although only e.g. 2/5 Analytes were selected in the source samples' multi-component analysis.

ramonski commented:
Just another thought that came into my mind:

To support all features of Analyses, e.g. LDL/UDL, Uncertainties, Text Result, Option fields etc., wouldn't it make sense to implement analytes as multi-references to existing Analysis Services?
Then we would not have the issue with the same keyword and all the other options that can be configured with analysis services.

xispa commented:
> Just another thought that came into my mind:
> 
> To support all features of Analyses, e.g. LDL/UDL, Uncertainties, Text Result, Option fields etc., wouldn't it make sense to implement analytes as multi-references to existing Analysis Services? Then we would not have the issue with the same keyword and all the other options that can be configured with analysis services.

This was my initial thought as well. However, note that there is a single analytical process involved with a multi-component analysis. I mean, the result of such analytical process is the concentrations of analytes/components, precisely. Thus, if they were services you'd need all them to always be configured the same way in most of the configuration settings, like units, interim fields, precision, etc. A multi-component service made of services that are barely similar among them won't make sense at all. 

Thus, I thought that instead of putting a lot of efforts to keep different services in-sync because they were assigned to the same multi-component (and also consider the possibility that might be unassigned later) it would be better and simpler if the multi-component service simply acted as the template of the analytes. This is precisely why the analytes are created as *Analysis* copies of the multi-component analysis they belong to (all analytes inherit same configuration from parent).

Although there are some settings (e.g. results ranges, MDL, etc.) for which we we will need to do some dance (something I wanted to do in another PR, btw), this copy approach I believe makes the things simpler. For instance, my idea re calculations (I saw you noticed this "weak point" already in one of the comments), was that you should be able to assign a calculation to a multi-component service. The calculation would not have any effect to the multi-component analysis though, but to the analytes only. This would be possible just by making the system to not render the interim fields in the analyses listing for multi-component analysis, but for analytes only. Another approach would be to just remove the calculation from the multi-component analysis as soon as the analytes are created. Imagine if you had different services, one per analyte, you should then to assign the same calculation to all them first, etc.. a nightmare imo.

Besides, a single multi-component service can comprehend a long list of analytes. For instance, for the determination of Volatile Organic Components (VOCs) one might define, quite easily, [more than 40](https://europepmc.org/article/med/12141544) or [60](https://www.labsertchemical.com/VOC-multi-component-standard-solutions/VOC-60-components-standard-005) different components/analytes. Also, not rare to have a multi-component analysis with 150 analytes. Imagine if the lab has say 4 different multi-component analyses, with that many components each! Cannot think an easy way to handle and keep all them in-sync.

So, these are the reasons why I decided to create copies of the analytes as *Analysis* instead of *AnalysisService*. But yes, I do know there is still work to do, specially regarding ranges and detection limits. I think I can add the calc thing in this PR though.

xispa commented:
> All Analytes are added when a sample is copied, although only e.g. 2/5 Analytes were selected in the source samples' multi-component analysis.

:sob:

ramonski commented:
Thanks for the explanation and I agree with the points you mentioned.
Maybe we can avoid the keyword clashes when we use the same validator for the analytes and maybe this approach for later:

We allow services to contain Analyte services that are created automatically when you add the rows in the multi component.

xispa commented:
> Thanks for the explanation and I agree with the points you mentioned. Maybe we can avoid the keyword clashes when we use the same validator for the analytes 

Yes, using same validator for analyte keywords as for services makes sense, yes.

> ... and maybe this approach for later:
> 
> We allow services to contain Analyte services that are created automatically when you add the rows in the multi component.

Not sure if I get the point. Even if we "hide" these "Analyte" services, we will always need to keep them up-to-date each time we do configuration changes to the multi-component service they belong to. This would mean to restore that relic of *ProxyField* or manually update the values of all analytes each time a value from a multi-component service is updated by adding a lot of redundant code in the setters. Don't thing this is more elegant than what we have with this "Analysis"-only approach. Maybe I am missing something though...

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Changelog

xispa committed:
Restore getRawMethod original implementation

xispa committed:
Fix test

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Remove unused function

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Fix F821 undefined name 'serviceapi'

xispa committed:
Do not create analytes for analytes on retest

xispa committed:
Do not apply (pre)Conditions to analytes

xispa committed:
Cleanup after_retest transition

xispa committed:
Fix doctest

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Default result of a multi-component is set to analytes

xispa committed:
Drag retests across analytes when retracting a multi-analysis

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Sort analytes as defined in the analysis servive

The analyes returned from samples are now sorted by title as per
https://github.com/senaite/senaite.core/pull/2196

This commit ensures the analytes are always returned in same
order as defined in the analysis service, and the multi component
is always returned as the lead analysis from its analytes

xispa committed:
Add leading zeros to the index used to sort analytes

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' into multi-analyses

xispa committed:
Merge branch '2.x' into multi-analyses

xispa committed:
Merge branch '2.x' into multi-analyses

xispa committed:
Merge branch '2.x' into multi-analyses

xispa committed:
Merge branch '2.x' into multi-analyses

xispa committed:
Fix max-depth recursion error on verify or retract

```
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 288, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module senaite.app.listing.view, line 236, in __call__
  Module senaite.app.listing.ajax, line 112, in handle_subpath
  Module senaite.core.decorators, line 22, in decorator
  Module senaite.app.listing.decorators, line 63, in wrapper
  Module senaite.app.listing.decorators, line 50, in wrapper
  Module senaite.app.listing.decorators, line 100, in wrapper
  Module senaite.app.listing.ajax, line 533, in ajax_transitions
  Module senaite.app.listing.ajax, line 225, in get_allowed_transitions_for
  Module senaite.app.listing.ajax, line 186, in get_transitions_for
  Module bika.lims.api, line 1076, in get_transitions_for
  Module Products.CMFPlone.WorkflowTool, line 107, in getTransitionsFor
  Module Products.DCWorkflow.DCWorkflow, line 395, in _checkTransitionGuard
  Module Products.DCWorkflow.Guard, line 93, in check
  Module Products.CMFCore.Expression, line 53, in __call__
  Module Products.PageTemplates.ZRPythonExpr, line 49, in __call__
   - __traceback_info__: here.guard_handler("verify")
  Module PythonExpr, line 1, in <module>
  Module Products.CMFCore.FSPythonScript, line 134, in __call__
  Module Shared.DC.Scripts.Bindings, line 335, in __call__
  Module Shared.DC.Scripts.Bindings, line 372, in _bindAndExec
  Module Products.PythonScripts.PythonScript, line 349, in _exec
  Module script, line 2, in guard_handler
   - <FSPythonScript at /senaite/guard_handler used for /senaite/clients/client-19/23AA047/ICP>
   - Line 2
  Module bika.lims.workflow, line 370, in guard_handler
  Module bika.lims.workflow.analysis.guards, line 83, in decorator
  Module bika.lims.workflow.analysis.guards, line 292, in guard_verify
  Module bika.lims.workflow.analysis.guards, line 547, in is_verified_or_verifiable
  Module bika.lims.workflow.analysis.guards, line 494, in is_transition_allowed
  Module bika.lims.workflow.analysis.guards, line 496, in is_transition_allowed
  Module plone.memoize.volatile, line 74, in replacement
  Module bika.lims.workflow.analysis.guards, line 527, in cached_is_transition_allowed
  Module bika.lims.workflow, line 233, in isTransitionAllowed
  Module Products.DCWorkflow.DCWorkflow, line 256, in isActionSupported
  Module Products.DCWorkflow.DCWorkflow, line 395, in _checkTransitionGuard
  ...
  Module Products.DCWorkflow.Guard, line 93, in check
  Module Products.CMFCore.Expression, line 53, in __call__
  Module Products.PageTemplates.ZRPythonExpr, line 49, in __call__
   - __traceback_info__: here.guard_handler("verify")
RuntimeError: maximum recursion depth exceeded in cmp
```

xispa committed:
Merge branch '2.x' into multi-analyses

xispa committed:
Merge branch '2.x' into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Remove unused imports

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Fix analytes in sample retests point to original multi component

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Fix doctest (seems sorting analyses does work differently)

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

xispa committed:
Compatibility with #2201

xispa committed:
Allow DL selector and Manual DL entry for Analytes

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into multi-analyses

ramonski committed:
Merge branch '2.x' into multi-analyses
